src/main/java/org/grohe/ondus/api/WebFormLogin.java

```diff
@@ -65,22 +72,45 @@ private String formTargetOf(String page) throws IOException {
 		}
 	}
 
-	private RefreshTokenResponse login(CloseableHttpClient httpclient, String actionUrl) throws IOException {
+	private RefreshTokenResponse login(CloseableHttpClient httpclient, String actionUrl) throws IOException, LoginException {
 		HttpPost post = new HttpPost(actionUrl);
 		post.setHeader("Content-Type", "application/x-www-form-urlencoded");
 		post.setHeader("X-Requested-With", "XMLHttpRequest");
 		post.setHeader("referer", actionUrl);
 		post.setHeader("origin", baseUrl);
-
 		post.setEntity(new StringEntity(buildLoginRequest()));
-
+		
 		try (CloseableHttpResponse response = httpclient.execute(post)) {
+			checkResponse(response);
+
 			return fetchToken(httpclient, fetchLocation(response));
 		}
 	}
 
-	private String fetchLocation(CloseableHttpResponse response) {
-		return response.getHeaders("Location")[0].getValue().replace("ondus://", "https://");
+	private String fetchLocation(CloseableHttpResponse response) throws LoginException, ParseException, IOException {
+		Header[] headers = response.getHeaders("Location");
+		
+		if (headers.length > 0) {
+			return headers[0].getValue().replace("ondus://", "https://");
+		}
+		else {
+			if (checkIsInvalidCredentialsResult(response)) {
```

Comment:
check can be removed, having "isInvalidResponse" is good enough regarding this. However, what about "isForbidden" instead of "invalid"? The result is not really invalid, it's just not the correct username/pasword :)